Upgrade to Hibernate ORM 5.2.6

Sanne opened the pull request:
This is for 5.7 only; we could backport the first two commits (HSEARCH-2549, HSEARCH-2550).

 - https://hibernate.atlassian.net/browse/HSEARCH-2380
 - https://hibernate.atlassian.net/browse/HSEARCH-2549
 - https://hibernate.atlassian.net/browse/HSEARCH-2550

Related:
 - https://hibernate.atlassian.net/browse/HHH-11309

Sanne commented:
contains commits from September & October last year :-/

Sanne commented on orm/src/test/java/org/hibernate/search/test/util/DdlTransactionIsolatorTestingImpl.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Copied from Hibernate ORM: being originally part of the `org.hibernate.test.util` package,
+ * this is not published in a jar which could be consumed by Hibernate Search.
+ * This copy is not meant to stay here for long.
```
I agreed with Steve that we should move the `DdlTransactionIsolatorTestingImpl.java` file to ORM testing utilities, so we'll be able to depend on it. Not least, it means they'll fix the file in case there's need for further changes ;)

So I prefer to use a plain copy of their class, to that now I know which class exactly I need to move.

Sanne commented on orm/src/main/java/org/hibernate/search/query/hibernate/impl/FullTextQueryImpl.java:
```diff
@@ -390,12 +390,12 @@ protected boolean isNativeQuery() {
 
 	@Override
-	public Query setEntity(int position, Object val) {
+	public org.hibernate.query.Query<R> setEntity(int position, Object val) {
```
Why would that be simpler?

I didn't put a great deal of thought here, as both alternatives are deprecated, and we don't even implement it ;)

yrodiere replied:
It would be simpler because `FullTextQuery` is what we want to implement in this file, while `org.hibernate.query.Query` is something we happen to inherit from indirectly. And that has the same name as another class we already reference in the same file.

Anyway, it's no important enough to argue about it... Let's keep it that way.

Sanne commented on integrationtest/spring/src/test/resources/beans.xml:
```diff
@@ -70,7 +70,7 @@
                 <entry key="hibernate.connection.isolation" value="2"/>
                 <!-- http://docs.jboss.org/hibernate/core/3.3/reference/en/html/architecture.html#architecture-current-session -->
-                <entry key="hibernate.current_session_context_class" value="org.hibernate.context.JTASessionContext"/>
+                <entry key="hibernate.current_session_context_class" value="org.hibernate.context.internal.JTASessionContext"/>
```
> Or is there a particular reason to reference the class explicitly?

Yes, me not knowing about the `jta` configuration option ;-)

Thanks, I'll change it.

Sanne commented on integrationtest/wildfly/pom.xml:
```diff
@@ -374,6 +374,39 @@
                             </resources>
                         </configuration>
                     </execution>
+                    <!-- Temporary workaround for HHH-11394 -->
```
right, thanks!

Sanne replied:
I didn't mean it as a strict rule, more like something to keep in mind.

New rule:
 - there never are strict rules.

I typically see value in keeping `git bisect` working fine; this is more valuable in maintenance branches as the likelyhood for it to be broken anyway during a large feature sprint is high.

There's also value in keeping changes indipendent. In this case I felt keeping each step of the ORM upgrade independent was making it clearer both for you to review, and for future reference.

If you feel strongly about it I can squash them all; what do you prefer?

yrodiere replied:
I don't mind either solution. I just thought you may have overlooked this potential issue with `git bisect`, but obviously you haven't. So let's keep it that way :)

Sanne commented:
@yrodiere thanks for the review. I addressed your remarks and added one more commit to solve 

 - https://hibernate.atlassian.net/browse/HSEARCH-2551

yrodiere commented on orm/src/main/java/org/hibernate/search/FullTextQuery.java:
```diff
@@ -13,7 +13,6 @@
 import org.apache.lucene.search.Sort;
 
 import org.hibernate.Criteria;
-import org.hibernate.Query;
```
It seems this import was being used in the javadocs, and the build now fails.
See lines 153 and 159.

Sanne replied:
damn :) I really need to figure out why such things don't fail to me locally.

yrodiere commented on orm/src/main/java/org/hibernate/search/FullTextQuery.java:
```diff
@@ -29,7 +28,7 @@
  * @author Hardy Ferentschik
  * @author Emmanuel Bernard
  */
-public interface FullTextQuery<R> extends Query<R>, ProjectionConstants, QueryImplementor<R> {
```
It's a small detail really, but since there's already an issue with this commit: `QueryImplementor` already extends `org.hibernate.query.Query` (and `org.hibernate.Query`, indirectly), so we could remove the explicit ` extends org.hibernate.query.Query<R>`. What do you think?

Sanne replied:
Interesting, I'm going to try that as it looks promising. FYI this wasn't possible in earlier versions of ORM 5.2.x, but those SPIs evolved a bit also based on our earlier attempts to migrate Search..

yrodiere commented:
Merged. Thanks!
